### PR TITLE
Homework 2 task 4 done

### PR DESCRIPTION
**Pull request by ivanzuev78**
Homework 2 task 4 done

### CODE REVIEW

File: HomeWork_2/test_hw4.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+    some_args = 100, 200
+
+    val_1 = cache_func(*some_args, s=90)
+    val_2 = cache_func(*some_args, s=90)
+
+    assert val_1 is val_2
+
+
+def test_cache_check_time():
```
Comment (Cjkjvfnby) [inline HomeWork_2/test_hw4.py:39]: Please add test for 

```python
def double(nums: List[int]) -> List[int]:
  return [x * 2 for x in nums]
```

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]

File: HomeWork_2/test_hw4.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+    cache_func = cache(func)
+
+    some_args = 100000, 200000
+
+    time_start = time.time()
```
Comment (Cjkjvfnby) [inline HomeWork_2/test_hw4.py:45]: Time is not a reliable source for test information.   See the task description for an assertion example.

Comment (ivanzuev78) [reply HomeWork_2/test_hw4.py:45]: We use cache to save time and computing power. But if our cache func is slower than original func, its a reason to think about caches realisation. 
I always have other tests like example.

File: HomeWork_2/test_hw4.py (hunk 1 above)
Comment (ivanzuev78) [reply HomeWork_2/test_hw4.py:39]: No problem.

File: HomeWork_2/hw4.py [hunk 3]
```diff
@@ -18,8 +18,21 @@ def func(a, b):
 assert val_1 is val_2
 
 """
-from collections import Callable
+from collections.abc import Callable
```
[the remark this answers is not in the document]
Comment (ivanzuev78) [reply HomeWork_2/hw4.py:21]: Done

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]

File: HomeWork_2/test_hw4.py (hunk 2 above)
Comment (ivanzuev78) [reply HomeWork_2/test_hw4.py:51]: I completely agree now.

File: HomeWork_2/test_hw4.py (hunk 1 above)
Comment (ivanzuev78) [reply HomeWork_2/test_hw4.py]: I had forgotten about hashability of some types. Now it should work correctly.

File: HomeWork_2/hw4.py [hunk 4]
```diff
@@ -18,8 +18,58 @@ def func(a, b):
...
+def object_is_cacheable(arg):
+
+    try:
+        hash(arg)
+        return True
+    except:
+        return False
+
+
+def hashable(users_input):
```
Comment (Cjkjvfnby) [inline HomeWork_2/hw4.py:33]: What is the difference between `object_is_cacheable` and `hashable`?

File: HomeWork_2/hw4.py [hunk 5]
```diff
@@ -18,8 +18,58 @@ def func(a, b):
...
 
 def cache(func: Callable) -> Callable:
-    ...
+
+    results = {}
+
+    def wrapper(*args, **kwargs):
+        if not hashable((*args, *kwargs)):
+            hash_ = make_it_hashable(*args, **kwargs)
+
```
Comment (Cjkjvfnby) [inline HomeWork_2/hw4.py:64]: This empty lines inside else blocks looks strange, can you remove them?

File: HomeWork_2/hw4.py [hunk 6]
```diff
@@ -18,8 +18,58 @@ def func(a, b):
...
+    return tuple(hashable_output)
 
 
 def cache(func: Callable) -> Callable:
-    ...
+
+    results = {}
+
+    def wrapper(*args, **kwargs):
+        if not hashable((*args, *kwargs)):
```
Comment (Cjkjvfnby) [inline HomeWork_2/hw4.py:62]: You can make something like this:
```python
key = make_key(*args, **kwargs)
if not key in results:
   results[key] = fun(*args, **kwargs)
return results[key]

```

File: HomeWork_2/hw4.py [hunk 7]
```diff
@@ -18,8 +18,58 @@ def func(a, b):
...
+
+
+def make_it_hashable(*args: Any, **kwargs: Any):
+    hashable_output = []
+    if all(map(hashable, args)):
+        hashable_output.append(tuple(i for i in args))
+    else:
+        for arg in args:
+            hashable_output.append(make_it_hashable(*arg))
+    if all(map(hashable, kwargs.values())):
```
Comment (Cjkjvfnby) [inline HomeWork_2/hw4.py:48]: You have lost keys.

please add test:  `foo(a=10) is foo(b=10)`

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]

File: HomeWork_2/hw4.py (hunk 4 above)
Comment (ivanzuev78) [reply HomeWork_2/hw4.py:33]: I'm so inattentive.

File: HomeWork_2/hw4.py (hunk 5 above)
Comment (ivanzuev78) [reply HomeWork_2/hw4.py:64]: Yes.

File: HomeWork_2/hw4.py (hunk 7 above)
Comment (ivanzuev78) [reply HomeWork_2/hw4.py:48]: Yes. Now its should be ok.

File: HomeWork_2/hw4.py (hunk 6 above)
Comment (ivanzuev78) [reply HomeWork_2/hw4.py:62]: Yes. It would be better.

File: HomeWork_2/hw4.py (hunk 7 above)
Comment (ivanzuev78) [reply HomeWork_2/hw4.py:48]: Added test "test_cache_same_args_different_kwargs"

Comment (ivanzuev78) [reply HomeWork_2/hw4.py:48]: Added test "test_cache_same_args_different_kwargs"

File: HomeWork_2/test_hw4.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+    some_args = 100, 200
+
+    val_1 = cache_func(*some_args, s=10)
+    val_2 = cache_func(*some_args, s=10)
+
+    assert val_1 is val_2
+
+
+def test_double_cache():
```
Comment (Cjkjvfnby) [inline HomeWork_2/test_hw4.py:57]: Funny name.

**Review by Cjkjvfnby — APPROVED** [review]

**Pull request merged**